### DISCUSSION TITLE
请问工具调用这块是怎么实现连续两次出现assistant?

### DISCUSSION THREAD

**Discussion by hongyix** [zai-org/ChatGLM3#546] [Q&A] [answered]
![image](https://github.com/THUDM/ChatGLM3/assets/14794327/05fc5ae6-bf19-4836-aca5-aff2ee2a5465)
这块请问是怎么实现的，不是规定assistant前面必须接user?
![image](https://github.com/THUDM/ChatGLM3/assets/14794327/a949b795-24a4-48fc-aadf-c76fd49b75b0)
实际在跑[cli_demo_tool.py](https://github.com/THUDM/ChatGLM3/blob/main/tool_using/cli_demo_tool.py) 的时候也没有成功复现两次assistant的情况。

**Comment by zRzRzRzRzRzRzR** [comment]
第一个assistant是Thought，很有可能不出现

**Reply by hongyix** [reply]
请问怎么区分第一个assistant结束后应该调用tools而不是等待user input呢？thoughts 和正常 assistant回答的区别是什么？

**Reply by minmie** [reply]
是啊，我之前看到这个也有疑问。
一般assistant之后要么是observation（通过tool_call判断） ，要么是user，如何判断何时应当连续进行两次assistant呢？
@zRzRzRzRzRzRzR

**Reply by zRzRzRzRzRzRzR** [reply]
意图识别目前6B是没法做到的，这条思维链是设计好的，一定会进tools

**Reply by MNYxin** [reply]
可以通过训练稳定出现2个assistant的回复吗？

**Reply by MNYxin** [reply]
如果微调不行的话，有什么替代的方案可以实现类似的效果吗

**Reply by zRzRzRzRzRzRzR** [reply]
暂时没有，不过你可以尝试langgraph等Agent框架，可能能行，但是还是不建议在这么小的模型尝试多个工具调用和连续调用了，建议使用API的大模型，成功率会高很多

**Reply by MNYxin** [reply]
在tool页签下，问一些不该调用工具的问题，也会随机选一个不相关的工具调用。这种情况有办法避免吗

**Reply by zRzRzRzRzRzRzR** [reply]
很正常，这个开源模型只能识别两三个工具就不行了

**Reply by MNYxin** [reply]
好的，谢谢

**Comment by hongyix** [comment]
![image](https://github.com/THUDM/ChatGLM3/assets/14794327/327eefc4-cf26-413d-9854-be416a0c8190)
我看了一下源码这块，不知道我理解的对不对。应该是第一个assistant的回答结尾不是<|user|> 或者<|observation|> 就接着输出第二个assistant。不过这又有一个新的问题，模型什么情况下会根据tokenizer.eos_token_id停止呢？这个对应 user 还是 observation呢？

**Reply by minmie** [reply]
哦，明白了，生成assistant的时候不会停止，其实是模型一次生成了2个assistant的回复，这是模型自己决定的。

**Reply by zRzRzRzRzRzRzR** [reply, marked as answer]
是的，第一个thoght就一定进去工具了
